File: slept/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>

int isLuhnValid(long long int number);
int cardt(long long int number);

int main()
{
    long long number =0 ;
    long long credit = get_long("Number: ");
    if (isLuhnValid(credit)) {
            cardt(credit);
        }
    else
    {
        printf("INVALID\n");
    }
}

int isLuhnValid(long long int number) {
    int sum = 0;
    int doubleDigit = 0;
    int i = 0;

    while (number > 0) {
        if (i % 2 == 0) {
            sum += number % 10;
        } else {
            doubleDigit = 2 * (number % 10);
            if (doubleDigit > 9) {
                sum += (doubleDigit % 10) + 1;
            } else {
                sum += doubleDigit;
            }
        }
        number /= 10;
        i++;
    }

    if (sum % 10 == 0) {
        return 1; // Valid
    } else {
        return 0; // Invalid
    }
}

int cardt(long long int number)
{
    int i = 0;
    int sum = 0;
    while (number > 0){
        if (number < 10)
        {
            i = number;
        }
        sum += 1;
        number /= 10;

    //printf("%d\n", i);
    }
    if ((sum == 15) && (i == 3))
    {
        printf("AMEX\n");
    }
    else if ((sum == 13) && (i == 4))
    {
        printf("VISA\n");
    }
    else if ((sum == 16) && (i == 4))
    {
        printf("VISA\n");
    }
    else if ((sum == 16) && (i == 5))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: slept/sentimental-credit/credit.py
#American Express uses 15-digit numbers \\ start with 34 or 37
#MasterCard uses 16-digit numbers \\ start with 51, 52, 53, 54, or 55
#Visa uses 13- and 16-digit numbers \\ start with 4

def check_sum(card):
    card = str(card)
    card = card[::-1]
    sum = 0
    for i in range(len(card)):
        if i % 2 == 0:
            sum += int(card[i])
        else:
            sum += int(card[i]) * 2
            if int(card[i]) * 2 > 9:
                sum -= 9
    if sum % 10 == 0:
        return True
    else:
        return False
card = input("Number: ")
if len(card) == 15:
    if card[0:2] == "34" or card[0:2] == "37":
        if check_sum(card):
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")
elif len(card) == 16:
    if card[0:2] == "51" or card[0:2] == "52" or card[0:2] == "53" or card[0:2] == "54" or card[0:2] == "55":
        if check_sum(card):
            print("MASTERCARD")
        else:
            print("INVALID")
    elif card[0] == "4":
        if check_sum(card):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
elif len(card) == 13:
    if card[0] == "4":
        if check_sum(card):
            print("VISA")
        else:
            print("INVALID")
    else:
            print("INVALID")
else:
    print("INVALID")
